Features/add services to domain layer

skaik-mo opened the pull request:
### Files Added
- Added `Task` data class and `TaskService` interface for task management.
- Added `Category` data class and `CategoryService` interface for category management.
- Created `Priority` and `State` enum classes to represent task priority and status.
- Added Exception types for Task and Category.
- Created `ExceptionHandler` class to map domain exceptions to user-friendly string resources.
- Implement `tryToExecute` suspending function for cleaner error handling in suspend functions.
- Added new string resources for various task and category error messages.
- Deleted the `usecase.kt` file from  usecase package as it was empty and served no purpose.

### Example usage tryToExecute
```
tryToExecute(
        action = { taskService.getTaskByID("TASK_ID") },
        onSuccess = {
            // handle success state
        },
        onError = { exception ->
            // handle error state
            ShowMessage(ExceptionHandler().mapExceptionToMessage(exception))
        }
    )
```

skaik-mo committed:
Add entiries in domain layer

- Added `Task` data class and `TaskService` interface for task management.
- Added `Category` data class and `CategoryService` interface for category management.
- Created `Priority` and `State` enum classes to represent task priority and status.

skaik-mo committed:
add Exception types for Task and Category

skaik-mo committed:
Remove `usecase.kt` as it is unnecessary.

This change deletes the `usecase.kt` file from the `com.qudus.tudee.domain.usecase` package as it was empty and served no purpose.

skaik-mo committed:
Add ExceptionHandler and support for localized error messages

- Create `ExceptionHandler` class to map domain exceptions to user-friendly string resources.
- Implement `tryToExecute` suspending function for cleaner error handling in suspend functions.
- Add new string resources for various task and category error messages.

Farah315 commented on app/src/main/java/com/qudus/tudee/domain/entity/Task.kt:
```diff
@@ -0,0 +1,15 @@
+    val title: String,
+    val description: String?,
+    val date: LocalDate,
+    val priority: Priority,
+    val state: State,
```
good work, but in my opinion we could improve the type name to make it specific like 
 ```
val priority: TaskPriority
val state: Task State
```

ElhamFadel replied:
Yes, also property date optional

NadeenRR replied:
The most accurate name is Priority, not TaskPriority, because the Priority inside the Task class is already clear and contextually follows the meaning of the task.

ElhamFadel replied:
@NadeenRR name enum class not the property

ismailElkalili replied:
I placed the enum class inside the Task class as an inner class

Farah315 commented on app/src/main/java/com/qudus/tudee/domain/repository/CategoryService.kt:
```diff
@@ -0,0 +1,11 @@
+
+import com.qudus.tudee.domain.entity.Category
+
+interface CategoryService {
```
since the category is part of the task, wouldn't it be better to name TaskCategoryService?

NadeenRR replied:
Same the last comment, the category Service is most correctly

Farah315 commented on app/src/main/java/com/qudus/tudee/domain/repository/CategoryService.kt:
```diff
@@ -0,0 +1,11 @@
+interface CategoryService {
+    suspend fun createCategory(category : Category)
+    suspend fun updateCategory(category : Category)
+    suspend fun deleteCategory(id: String)
```
why the type to id is string? in our data class is Long,
 and id , it's better to change the name like categoryId, It was a little confusing

NadeenRR replied:
Yes skaik, edit the id to long

Farah315 commented on app/src/main/java/com/qudus/tudee/domain/repository/CategoryService.kt:
```diff
@@ -0,0 +1,11 @@
+    suspend fun updateCategory(category : Category)
+    suspend fun deleteCategory(id: String)
+    suspend fun getCategories(): List<Category>
+    suspend fun getCategoryById(id: String): Category
```
Same thing for id type

Farah315 commented on app/src/main/java/com/qudus/tudee/domain/repository/TaskService.kt:
```diff
@@ -0,0 +1,14 @@
+
+interface TaskService {
+    suspend fun createTake(task: Task)
+    suspend fun updateTask(task: Task)
```
Is it good to add for example
    suspend fun editTaskSatatus(state: State): List<Task> ?
Since there getTaskbyStatus?

ismailElkalili replied:
we make moveToState  it is same idea(editTaskSatatus) but we naming like this becaues we have button just moving to next state

Farah315 commented on app/src/main/java/com/qudus/tudee/domain/type/Priority.kt:
```diff
@@ -0,0 +1,7 @@
+package com.qudus.tudee.domain.type
+
+enum class Priority {
```
What about rename it to Takepriority?

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/domain/exception/CategoryException.kt:
```diff
@@ -0,0 +1,18 @@
+package com.qudus.tudee.domain.exception
+
+open class CategoryException : Exception()
```
Since we’re building a custom exception hierarchy, I recommend introducing a base exception class like `TudeeException : Exception()`. Then, make `TaskException` and `CategoryException` **extend** from `TudeeException`. This way, all exceptions in our app will be part of a unified exception hierarchy, which improves consistency and makes it easier to handle all domain-level errors in one place when needed.

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/ui/exception/ExceptionHandler.kt:
```diff
@@ -0,0 +1,71 @@
+            else -> handleUnexpectedException(exception = exception)
+        }
+    }
+
+    @Composable
+    private fun handleTaskException(exception: Throwable): String {
+        return when (exception) {
+            is TaskCreationFailedException -> stringResource(R.string.failed_to_create_task)
+            is TaskUpdateFailedException -> stringResource(R.string.failed_to_update_task)
+            is TaskDeletionFailedException -> stringResource(R.string.failed_to_delete_task)
+            is TaskFetchAllFailedException -> stringResource(R.string.failed_to_fetch_tasks)
+            is TaskNotFoundException -> stringResource(R.string.no_tasks_found)
+            is TaskReadFailedException -> stringResource(R.string.failed_to_read_task)
+            is EmptyTaskTitleException -> stringResource(R.string.task_title_cannot_be_empty)
+            is TaskStateChangeFailedException -> stringResource(R.string.failed_to_change_task_state)
+            else -> handleUnexpectedException(stringResource(R.string.task_error), exception)
+        }
+    }
+
+    @Composable
+    private fun handleCategoryException(exception: Throwable): String {
+        return when (exception) {
+            is CategoryCreationFailedException -> stringResource(R.string.failed_to_create_category)
+            is CategoryUpdateFailedException -> stringResource(R.string.failed_to_update_category)
+            is CategoryDeletionFailedException -> stringResource(R.string.failed_to_delete_category)
+            is CategoryFetchAllFailedException -> stringResource(R.string.failed_to_fetch_categories)
+            is CategoryNotFoundException -> stringResource(R.string.no_categories_found)
+            is CategoryReadFailedException -> stringResource(R.string.failed_to_read_category)
+            is EmptyCategoryTitleException -> stringResource(R.string.category_title_cannot_be_empty)
+            else -> handleUnexpectedException(stringResource(R.string.category_error), exception)
+        }
+    }
+
+    @Composable
+    private fun handleUnexpectedException(
+        exceptionType: String = stringResource(R.string.unexpected_error),
+        exception: Throwable
+    ): String {
+        return "${exceptionType}: ${exception.message ?: stringResource(R.string.unexpected_error)}"
+    }
```
If we unify all exceptions under a single base class like `TudeeException`, we can simplify this logic by using just **one mapping function** (e.g., handleTudeeException) instead of having separate ones for TaskException and CategoryException. what do you see ?

Farah315 reviewed (commented):
great job,I'll be happy when we discuss exceptions

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/ui/exception/tryToExecute.kt:
```diff
@@ -0,0 +1,16 @@
+package com.qudus.tudee.ui.exception
+
+suspend fun <T> tryToExecute(
+    action: suspend () -> T,
+    onSuccess: suspend (T) -> Unit = {},
+    onError: suspend (Throwable) -> Unit = {},
+    completion: () -> Unit = {}
+) {
+    try {
+        onSuccess(action())
+    } catch (exception: Throwable) {
+        onError(exception)
+    } finally {
+        completion()
+    }
+}
```
I don’t think this function belongs in the ui.exception package. Since we’ll be using `tryToExecute `primarily in **view models**, it would make more sense to move it into a `BaseViewModel `. That way, it’s scoped closer to where it's used and better aligned with MVVM responsibilities.

Also, implementing it directly in `BaseViewModel `allows us to encapsulate `viewModelScope.launch(...)` with a default `Dispatcher`, which helps avoid repeating boilerplate every time we need to call a suspend function. This keeps our code cleaner and more consistent across all view models.

### **My Suggestion:**
```
fun <T> tryToExecute(
    action: suspend () -> T,
    onSuccess: (T) -> Unit = {},
    onError: suspend (TudeeException) -> Unit = {},
    completion: () -> Unit = {},
    dispatcher: CoroutineDispatcher = Dispatchers.IO
) {
    viewModelScope.launch(dispatcher) {
        try {
            onSuccess(action())
        } catch (exception: Throwable) {
            onError(exception as? TudeeException ?: TudeeException())
        } finally {
            completion()
        }
    }
}

```
This way, we make the exception handling and coroutine usage centralized and reusable.

Belal-Alkhatib reviewed (changes requested):
You did a great job here @skaik-mo 👏

ElhamFadel commented on app/src/main/java/com/qudus/tudee/domain/type/State.kt:
```diff
@@ -0,0 +1,7 @@
+package com.qudus.tudee.domain.type
+
+enum class State {
```
State of what? we need to rename this to TaskState

Mohammed-qmr replied:
The name State is suitable because it represents the task’s status, but the important point here is that the name is very general and may cause confusion when there are other types of states in the project, such as UI state or system state.

ElhamFadel commented on app/src/main/java/com/qudus/tudee/domain/entity/Task.kt:
```diff
@@ -0,0 +1,15 @@
+    val id: Long,
+    val title: String,
+    val description: String?,
+    val date: LocalDate,
```
date of what ? createdAt should be more readable

Mohammed-qmr commented on app/src/main/java/com/qudus/tudee/domain/repository/TaskService.kt:
```diff
@@ -0,0 +1,14 @@
+    suspend fun createTake(task: Task)
+    suspend fun updateTask(task: Task)
+    suspend fun deleteTask(id: String)
+    suspend fun getTasks(id: String): List<Task>
```
The parameter id in the function getTasks(id: String): List<Task> is unclear. Could you please clarify what this id represents?
Is it intended to filter tasks by category, user, or something else?
If the function is meant to fetch all tasks without any filtering, it might be better to remove this parameter for clarity.

Also, I noticed that the id type is sometimes String and sometimes Long in other parts of the code. Could you please clarify which type should be used consistently, and why there is this difference?

ismailElkalili committed:
Refactor PR commit:
- Remove Priority and State class to entity
- make TudeeExecption and make him extend Execption and make

ismailElkalili committed:
adding libs.kotlinx.datetime to dependencies

ismailElkalili committed:
Merge remote-tracking branch 'origin/develop' into features/add-services-to-domain-layer

# Conflicts:
#	app/src/main/java/com/qudus/tudee/ui/designSystem/component/BottomNavBar.kt
#	app/src/main/java/com/qudus/tudee/ui/designSystem/component/CategoryTaskComponent.kt
#	app/src/main/java/com/qudus/tudee/ui/designSystem/component/DatePicker.kt
#	app/src/main/java/com/qudus/tudee/ui/designSystem/component/PriorityLevel.kt
#	app/src/main/java/com/qudus/tudee/ui/designSystem/component/SnackBar.kt
#	app/src/main/java/com/qudus/tudee/ui/designSystem/component/TabBar.kt
#	app/src/main/java/com/qudus/tudee/ui/designSystem/component/TaskTabItem.kt
#	app/src/main/java/com/qudus/tudee/ui/designSystem/component/TudeeChip.kt
#	app/src/main/java/com/qudus/tudee/ui/designSystem/theme/TudeeTheme.kt
#	app/src/main/res/values/strings.xml

ismailElkalili committed:
fix: solve some conflict
